static/app/views/dashboardsV2/datasetConfig/context.tsx

```diff
@@ -0,0 +1,127 @@
+import {ReactNode, useState} from 'react';
+
+import {
+  EventsStats,
+  Group,
+  MetricsApiResponse,
+  MultiSeriesEventsStats,
+  SessionApiResponse,
+} from 'sentry/types';
+import {EventsTableData, TableData} from 'sentry/utils/discover/discoverQuery';
+import {createDefinedContext} from 'sentry/utils/performance/contexts/utils';
+
+import {WidgetType} from '../types';
+
+import {DatasetConfig} from './base';
+
+interface DatasetConfigContextInterface<SeriesResponse, TableResponse> {
+  setDatasetConfig: (value?: DatasetConfig<SeriesResponse, TableResponse>) => void;
+  datasetConfig?: DatasetConfig<SeriesResponse, TableResponse>;
+}
+
+function createDatasetConfigContext<SeriesResponse, TableResponse>() {
+  const [_DatasetConfigProvider, useDatasetConfigContext, DatasetConfigContext] =
+    createDefinedContext<DatasetConfigContextInterface<SeriesResponse, TableResponse>>({
+      name: 'DatasetConfigContext',
+    });
+
+  function DatasetConfigProvider({children}: {children: ReactNode}) {
+    const [datasetConfig, setDatasetConfig] = useState<
```

Comment:
hmmm just thinking about this a little bit. Could we get away with not using `useState`?
In theory, I _think_ we only need to set the `datasetConfig` for each data set once in the app.
If we only need to set it once, maybe we can just pass it as an arg to `createDatasetConfigContext` when creating each of the dataset config Contexts? That way, we don't need to create a consumer to call `setDatasetConfig`.

I could be missing the reason why we do this though. Let me know!

Reply:
Agreed. It would be really great to have the context initialize it from the beginning. I've also been messing around with this but I've been having some challenges with the concrete types in the config and the generic types being used in other parts of this util

Reply:
Addressed Ed's comments and implemented Nar's suggestion to pass widgetType to createDatasetConfigContext!

Reply:
Addressed Ed's comments and implemented Nar's suggestion to pass widgetType to createDatasetConfigContext!